File: javascript/problemsets/datastructures/SinglyLinkedList.js

class SinglyLinkedList {

    constructor() {
        this.head = null;
        this.count = 0;
    }

    addToFront(value) {
        let newNode = new Node(value);
        if (this.head == null) {
            this.head = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }
        this.count++;
    }

    /**
     * @return {Array} 
     */
    toArray() {
        let curr = this.head;
        let returnArr = [];
        let currCount = 0;
        while(curr != null) {
            returnArr.push(curr.data);
            curr = curr.next;
            if (currCount > this.count) {
                throw new Error("List is corupted");
            }
            currCount++;
        }
        return returnArr;
    }

    getNodeAt(index) {
        let curr = this.head;
        let currIndex = 0;
        while (curr != null) {
            if (currIndex === index) {
                return curr;
            }
            curr = curr.next;
            currIndex++;
        }

        return curr;
    }

    //2.5
    detectCycle() {
        let slowPtr = this.head;
        let fastPtr = this.head.next;
        while (slowPtr != null && fastPtr != null) {
            if (slowPtr === fastPtr) {
                return true;
            }
            slowPtr = slowPtr.next;
            fastPtr = fastPtr.next.next;
        }
        return false;
    }

    detectCycleKeyTable() {
        let seenNodes = {};
        let curr = this.head;
        while (curr != null) {
            if (seenNodes[curr.uniqueId]) {
                return true;
            }
            seenNodes[curr.uniqueId] = true;
            curr = curr.next;
        }
        return false;
    }
}

class Node {

    constructor(data) {
        this.data = data;
        this.next = null;
        this.uniqueId = guid();
    }
}

function guid() {
    function s4() {
      return Math.floor((1 + Math.random()) * 0x10000)
        .toString(16)
        .substring(1);
    }
    return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4();
  }

export { SinglyLinkedList };

File: java/src/com/datastructures/linkedlists/SinglyLinkedList.java
package com.datastructures.linkedlists;

import com.datastructures.Node;

public class SinglyLinkedList<T> {

    private Node<T> head = null;
    private int count;
    public SinglyLinkedList() {
        this.count = 0;
    }

    public SinglyLinkedList(T value) {
        this.add(value);
    }

    public void add(T value) {
        Node<T> newHead = new Node<T>(value);
        newHead.setNext(head);
        head = newHead;
        this.count++;
    }

    public void add(T value, int index) {
        
        if(index > count) {
            throw new IndexOutOfBoundsException();
        }

        if(index == 0) {
            add(value);
            return;
        }

        Node<T> curr = head;
        Node<T> newNode = new Node<T>(value);
        int currIndex = 0;
        while(curr != null) {
            if(currIndex == index-1) {
                newNode.setNext(curr.getNext());
                curr.setNext(newNode);
                break;
            }
            if(curr.getNext() == null) {
                curr.setNext(newNode);
                break;
            }
            curr = curr.getNext();
        }
        this.count++;
    }

    @Override
    public String toString() {
        Node<T> curr = head;
        String returnString = "";
        while(curr != null) {
            returnString += (curr.getData() + "->");
            curr = curr.getNext();
        }
        returnString +="null";
        return returnString;
    }
}